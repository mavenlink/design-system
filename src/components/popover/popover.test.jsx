import React, { useRef, Fragment } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Popover from './popover.jsx';

describe('Popover', () => {
  function PopoverWithToggle(popoverProps) {
    const popoverRef = useRef();

    return (
      <Fragment>
        <button onClick={() => { popoverRef.current.open = true; }}>Open Popover</button>
        <button onClick={() => { popoverRef.current.open = false; }}>Close Popover</button>
        <span>Some text</span>
        <Popover ref={popoverRef} {...popoverProps}>{popoverProps.children}</Popover>
      </Fragment>
    );
  }

  it('has defaults', () => {
    const popoverRef = React.createRef();

    render(<Popover title="Defaults" />);
    expect(document.body).toMatchSnapshot();

    render(<Popover title="Defaults" startOpen={true} />);
    expect(document.body).toMatchSnapshot();

    render(<Popover ref={popoverRef} title="Defaults with ref" />);
    expect(popoverRef.current).toMatchSnapshot();
  });

  it('has a title', () => {
    render(<PopoverWithToggle title="Another title" />);

    userEvent.click(screen.getByText('Open Popover'));
    expect(screen.getByRole('dialog', { name: 'Another title' })).toBeInTheDocument();
  });

  it('renders children', () => {
    render(<PopoverWithToggle title="Another title"><span>More text</span></PopoverWithToggle>);

    userEvent.click(screen.getByText('Open Popover'));
    expect(screen.getByText('More text'));
  });

  it('can be toggled open/closed', () => {
    render(<PopoverWithToggle title="Another title" />);

    userEvent.click(screen.getByText('Open Popover'));
    expect(screen.getByText('Another title')).toBeInTheDocument();

    userEvent.click(screen.getByText('Close Popover'));
    expect(screen.queryByText('Another title')).not.toBeInTheDocument();
  });

  it('can be closed by clicking elsewhere in the window', () => {
    render(<PopoverWithToggle title="Another title" />);

    userEvent.click(screen.getByText('Open Popover'));
    expect(screen.getByText('Another title')).toBeInTheDocument();

    userEvent.click(screen.getByText('Some text'));
    expect(screen.queryByText('Another title')).not.toBeInTheDocument();
  });

  it('can be closed by focusing outside', () => {
    render(<PopoverWithToggle title="Another title" />);

    userEvent.click(screen.getByText('Open Popover'));

    userEvent.tab();
    userEvent.tab();
    expect(screen.getByText('Open Popover')).toHaveFocus();
    expect(screen.queryByText('Another title')).not.toBeInTheDocument();
  });

  it('opens with focus on the close iconbutton', () => {
    render(<PopoverWithToggle title="Another title" />);

    userEvent.click(screen.getByText('Open Popover'));
    expect(document.activeElement).toContainElement(screen.getByTitle('Close popover'));
  });

  it('can be closed by activating the close iconbutton', () => {
    render(<PopoverWithToggle title="Another title" />);

    userEvent.click(screen.getByText('Open Popover'));
    userEvent.click(document.activeElement);
    expect(screen.queryByText('Another title')).toEqual(null);

    userEvent.click(screen.getByText('Open Popover'));
    fireEvent.keyDown(document.activeElement, { key: 'Spacebar' });
    expect(screen.queryByText('Another title')).toEqual(null);

    userEvent.click(screen.getByText('Open Popover'));
    fireEvent.keyDown(document.activeElement, { key: 'Enter' });
    expect(screen.queryByText('Another title')).toEqual(null);
  });

  it('does not close when clicking within itself', () => {
    render(<PopoverWithToggle title="Another title"><span>More text</span></PopoverWithToggle>);

    userEvent.click(screen.getByText('Open Popover'));
    userEvent.click(screen.getByText('More text'));
    expect(screen.getByText('Another title').tagName).toEqual('H1');
  });

  describe('flush api', () => {
    it('can flush left', () => {
      render(<PopoverWithToggle title="Another title" flush="left"><span>More text</span></PopoverWithToggle>);
      userEvent.click(screen.getByText('Open Popover'));
      expect(screen.getByRole('dialog')).toHaveStyle({ left: 0 });
      expect(screen.getByRole('dialog')).not.toHaveStyle({ right: 0 });
    });

    it('can flush right', () => {
      render(<PopoverWithToggle title="Another title" flush="right"><span>More text</span></PopoverWithToggle>);
      userEvent.click(screen.getByText('Open Popover'));
      expect(screen.getByRole('dialog')).toHaveStyle({ right: 0 });
      expect(screen.getByRole('dialog')).not.toHaveStyle({ left: 0 });
    });
  });
});
