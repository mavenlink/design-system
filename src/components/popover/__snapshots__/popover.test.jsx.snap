// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover has defaults 1`] = `
<body>
  <div />
</body>
`;

exports[`Popover has defaults 2`] = `
<body>
  <div />
  <div>
    <section
      aria-labelledby="popover-heading"
      class="container"
      role="dialog"
    >
      <div
        role="presentation"
      >
        <div
          class="heading-container"
          id="popover-heading"
        >
          <h1
            class="heading"
          >
            Defaults
          </h1>
          <svg
            class="close-button"
            height="8"
            role="button"
            tabindex="0"
            width="8"
          >
            <title>
              Close popover
            </title>
            <use
              xlink:href="#clear.svg"
            />
          </svg>
        </div>
      </div>
    </section>
  </div>
</body>
`;

exports[`Popover has defaults 3`] = `
Object {
  "open": false,
}
`;
